nixos/public-inbox: enable and test confinement

ju1m opened the pull request:
## Description of changes
- [X] ~Fix typo `BindPathsReadOnly` => `BindReadOnlyPaths`~ (was fixed in 14c57ce7f7c207072ed895c2fad06ef2b2ca1ca3) and actually test the binding with `gitolite` repositories.
- [X] Enable `ProtectSystem = "strict"`.
- [X] Enable `confinement`.
- [X] Test `confinement`.
- [X] Test `systemd-analyze security` thresholds.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ju1m commented:
- [X] Rebase after the merge of https://github.com/NixOS/nixpkgs/pull/289593

ju1m commented:
- [X] Fix typo `InacessiblePaths` => `InaccessiblePaths`

ju1m commented:
- [X] Rebase after:
  - 4f0dadbf38ee4cf4cc38cbc232b7708fddf965bc 'treewide: format all inactive Nix files'
  - 6b53949b0cbe3f6dadf4ea730558ac5f4eb56793 "formats.ini: expose INI atom from all ini formats"

alyssais commented:
Sorry I've taken so long with this.  Would you mind splitting it into multiple commits?  At least the test changes and module changes should be ungrouped.

ju1m committed:
nixos/public-inbox: enable confinement

ju1m commented:
@alyssais, thanks for following this PR. I've split into two commits (module and test) and simplified a lot the changes by just enabling `confinement` by default, as done in `radicle` and `cryptpad`.

alyssais commented on nixos/modules/services/mail/public-inbox.nix:
```diff
@@ -126,28 +117,9 @@ let
       confinement = {
-        # Until we agree upon doing it directly here in NixOS
-        # https://github.com/NixOS/nixpkgs/pull/104457#issuecomment-1115768447
-        # let the user choose to enable the confinement with:
```
Are there now other modules that use confinement by default?

ju1m replied:
@alyssais, since https://github.com/NixOS/nixpkgs/pull/289593 has been merged last May, those modules enable `confinement` by default:
- [radicle](https://github.com/NixOS/nixpkgs/blob/ab6998e83282a01827f660e5e25ea526ac559a50/nixos/modules/services/misc/radicle.nix#L103) (co-maintained by me)
- [cryptpad](https://github.com/NixOS/nixpkgs/blob/ab6998e83282a01827f660e5e25ea526ac559a50/nixos/modules/services/web-apps/cryptpad.nix#L209) (not co-maintained by me)

alyssais replied:
Thanks! OOC, in https://github.com/NixOS/nixpkgs/pull/234641 (which maybe I should finish), I noted that DynamicUser was not compatible with confinement.  Do you know if that's changed?

ju1m replied:
@alyssais, yes in https://github.com/NixOS/nixpkgs/pull/289593 we made `confinement` compatible with `ProtectSystem=strict` (and thus with `DynamicUser=1` too). I'm following https://github.com/NixOS/nixpkgs/pull/234641 and will review it, should you have time to finish it.

alyssais replied:
Great, thanks :)

alyssais commented on nixos/tests/public-inbox.nix:
```diff
@@ -189,12 +200,37 @@ import ./make-test-python.nix (
+                  '') gitoliteRepositories
+                )
+              }",
+              "(cd gitolite-admin && git add . && git commit -m 'Add repositories' && git push)",
```
Why is this being done over SSH on the same machine?

ju1m replied:
Cloning `gitolite-admin.git` and pushing configurations through SSH is the normal way to control Gitolite, that's [why a pubkey is required](https://github.com/NixOS/nixpkgs/blob/309e8a5e50b782930986a23f41bc91186cac38c6/nixos/modules/services/misc/gitolite.nix#L245) (well, unless `repositories` already exists, but I think it's better to not rely on the internals of the module, and IMHO it's simpler to use SSH).

alyssais commented:
Looks like https://github.com/NixOS/nixpkgs/pull/308740 broke the NixOS test, so I think we should probably fix that without module changes first.

ju1m commented:
@alyssais, sorry, where do you see that the NixOS test is broken?

alyssais commented:
[e.g.](https://hydra.nixos.org/build/283284731)

ju1m commented:
> [e.g.](https://hydra.nixos.org/build/283284731)

This is due to `BindReadOnlyPaths=` to non-existing Git repositories:
```
machine # [   18.093339] (ox-httpd)[1492]: public-inbox-httpd.service: Failed to set up mount namespacing: /var/lib/gitolite/repositories/user/repo1.git: No such file or directory
machine # [   18.095725] (ox-httpd)[1492]: public-inbox-httpd.service: Failed at step NAMESPACE spawning /nix/store/58ywrdv08k1gjahcx9plv2makzk1kqjq-perl5.40.0-public-inbox-1.9.0/bin/public-inbox-httpd: No such file or directory
```
So https://github.com/NixOS/nixpkgs/pull/308740 was merged with a failing test, and now there is no trace of this unless diving into hydra logs? That's quite unfortunate.
This PR configures `gitolite` in the test to avoid this error.

alyssais commented:
> So #308740 was merged with a failing test, and now there is no trace of this unless diving into hydra logs? That's quite unfortunate.

Yeah.  OfBorg has never known how to test changes to NixOS modules to my knowledge, so reviewers have to remember to tell it to.

> This PR configures `gitolite` in the test to avoid this error.

Great! That's not at all obvious from the commits though.  Could the test be fixed first, in its own commit (ideally its own PR)?  That way it's easier to understand what's happened, and also to avoid breaking the test again if e.g. the rest of this needs to be reverted.

ju1m commented:
@ofborg test public-inbox

alyssais commented on nixos/tests/public-inbox.nix:
```diff
@@ -10,7 +10,6 @@ import ./make-test-python.nix (
       install -D -t $out key.pem cert.pem
     '';
 
-    # Git repositories paths in Gitolite.
```
Is this removal intentional?

alyssais replied:
Ah okay, thanks! Should probably be a separate commit since it's not related to testing confinement.

ju1m committed:
nixos/public-inbox: test confinement

ju1m committed:
nixos/public-inbox: cleanup test

alyssais commented:
@OfBorg test public-inbox

alyssais commented:
Thanks for your persistence! I'm sorry this took so long.
